test/suite/test_cursor_bound19.py

```diff
@@ -0,0 +1,270 @@
+            create_params += "k{0},".format(str(start)) 
+            start += 1
+
+        start = 0
+        for _ in self.value_format:
+            create_params += "v{0},".format(str(start)) 
+            start += 1
+        create_params += ")"
+
+        if (self.use_colgroup):
+            create_params += ",colgroups=("
+            start = 0
+            for _ in self.value_format:
+                create_params += "g{0},".format(str(start)) 
+                start += 1
+            create_params += ")"
+        return create_params
+
+    def check_val(self, i):
+        list_key = []
+        for key in self.value_format:
+            if key == 'S':
+                list_key.append(str(i))
+            elif key == "r":
+                list_key.append(self.recno(i))
+            elif key == "i":
+                list_key.append(i)
+            elif key == "u":
+                list_key.append(str(i).encode())
+        
+        if (len(self.value_format) == 1):
+            return list_key[0]
+        else:
+            return list_key
+
+    def cursor_traversal_bound(self, cursor, lower_key, upper_key, next=None, expected_count=None):
+        if next == None:
+            next = self.direction
+
+        start_range = self.start_key
+        end_range = self.end_key
+
+        if (upper_key):
+            if (self.gen_val(upper_key) < end_range):
+                end_range = upper_key
+                
+        if (lower_key):
+            if (self.gen_val(lower_key) > start_range):
+                start_range = lower_key
+
+        count = ret = 0
+        while True:
+            if (next):
+                ret = cursor.next()
+            else:
+                ret = cursor.prev()
+            self.assertTrue(ret == 0 or ret == wiredtiger.WT_NOTFOUND)
+            if ret == wiredtiger.WT_NOTFOUND:
+                break
+            count += 1
+            key = cursor.get_key()
+            
+            if (lower_key):
+                self.assertTrue(self.check_val(lower_key) <= key)
+                
+            if (upper_key):
+                self.assertTrue(key <= self.check_val(upper_key))
+            
+        if (expected_count != None):
+            self.assertEqual(expected_count, count)
+        else:
+            self.assertEqual(end_range - start_range + 1, count)
+
+    def create_session_and_cursor(self, cursor_config=None):
+        uri = self.uri + self.file_name
+        create_params = 'value_format={},key_format={}'.format(self.value_format, self.key_format)
+        create_params += self.gen_index_create_param()
+        self.session.create(uri, create_params)
+
+        # Add in column group.
+        if self.use_colgroup:
+            for i in range(0, len(self.value_format)):
+                create_params = 'columns=(v{0}),'.format(i)
+                suburi = 'colgroup:{0}:g{1}'.format(self.file_name, i)
+                self.session.create(suburi, create_params)
+
```

Comment:
Yeah the original code for this is in wtbounds.py, but the biggest problem is that I wanted the populated values to be duplicates of each other to test index tables specifically. Do you think we should extract the populate side of the function out instead? I am more inclined to do this as an improvement feature.

Reply:
Any code that looks the same or substantially the same should be refactored into helper methods, or base classes, or some other approach that leaves you with a single copy.  I'm not saying that you always need to take this entire function and make it a common function.  However, what I was pointing out was that lines 160-165 are exactly the same as test_cursor_bound02.py: line 75.  And test_cursor_bound02.py:147 (so I suspect there are other occurrences, possibly many).  So those lines at least, should be refactored in some way.  It really helps readability and maintainability!  I would at least (in this commit), create a common function for this particular idiom and start using it going forward.  It would be fine to file an improvement ticket to use it elsewhere.

Personally, any time I have the temptation to copy any code, I first think about whether it belongs in a shared method.

Reply:
Done! i have merged it inside wtbounds